src/main/java/com/openlattice/authorization/HazelcastAuthorizationService.java

```diff
@@ -440,55 +430,57 @@ public void deletePermissions( AclKey aclKey ) {
     public Map<AclKey, EnumMap<Permission, Boolean>> authorize(
             Map<AclKey, EnumSet<Permission>> requests,
             Set<Principal> principals ) {
-        AuthorizationAggregator agg = aces
-                .aggregate( new AuthorizationAggregator(
-                                transformValues( requests, HazelcastAuthorizationService::noAccess ) ),
-                        matches( requests.keySet(), principals ) );
 
-        return agg.getPermissions();
+        Map<AclKey, EnumMap<Permission, Boolean>> permissionMap = Maps.newHashMap( transformValues( requests,
+                HazelcastAuthorizationService::noAccess ) );
+
+        aces.executeOnEntries( new AuthorizationEntryProcessor(), matches( requests.keySet(), principals ) )
```

Comment:
It turns out executeOnEntries doesn't properly hold locks (there are hazelcast issues about this and we had issues with id generation as a result).  The correct thing to do here is to generate all AceKey and then only execute on those keys using an Offloadable and ReadOnlyEntryProcessor. This will also avoid executing each authorization queries on all nodes, which is good for scalability.

Reply:
so basically switch to executeOnKeys instead of executeOnEntries? Or are you saying the entry processor itself should be changed in some way? (AuthorizationEntryProcessor is already Offloadable and ReadOnly)

Reply:
Switch to executeOnKeys instead of executeOnEntries, no changes required to entry processor. The catch is that it will require serialization of a large number of keys instead of just the entry processor. We will want to have before and after timing numbers if possible to see if it hurts performance (i.e we may see a perf hit from serializing the keys, but it may scale better in the long wrong).

Reply:
I made that change :)

Reply:
oh. you merged. lol